cpp/src/rolling/range_window_bounds.cpp

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/rolling/range_window_bounds.hpp>
+#include <cudf/scalar/scalar_factories.hpp>
+#include <cudf/types.hpp>
+#include <cudf/wrappers/durations.hpp>
+#include "range_window_bounds_detail.hpp"
+
+namespace cudf {
+namespace {
+
+struct range_scalar_constructor {
+  template <typename T, std::enable_if_t<!detail::is_supported_range_type<T>(), void>* = nullptr>
+  scalar* operator()(scalar const& range_scalar_) const
```

Comment:
It's preferred to have factory methods return unique_ptr. This makes the ownership transfer explicit. Similarly, range_window_bounds should take a unique_ptr since it's a sink for the scalar.

Reply:
>  (it's a bit scary to see a new call, especially without any delete calls).

Hmm. I do see your point about this being not being transparent. I'll check if I can restructure this a little better, and try do away with the raw pointer.

Reply:
I have switched the factory implementation to return `unique_ptr`, which I then `std::move()` to a `shared_ptr`. Thank you for this suggestion.